### PR TITLE
Add more examples of cylc workflows and ebseg – for discussion

### PR DESCRIPTION
**Pull request by hollandjg**
Add some example workflows demonstrating different ways of running `ebseg` under cylc. 

- `workflow/cylc/beaufort-sea-hudson-bay`: run the analysis on two large regions in Beaufort Sea and Hudson Bay
- `workflow/cylc/paper-domain`: run the analysis on the regions and the time-periods shown in the Buckley paper
- `workflow/cylc/sampled-examples`: run the analysis on the regions and the time-periods from https://github.com/danielmwatkins/ift_cal-val/tree/main/data/ift_data/ift_case_definitions. This also uses `cylc-rose` to treat the workflow as an executable, and has some examples of utility scripts of the kind we might want to use to make the workflow code simpler. 

These are small examples, but might be useful for testing different approaches before we start refactoring the main IFT pipeline.

### CODE REVIEW

File: workflow/cylc/beaufort-sea-hudson-bay/README.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
+## Cylc
+To run the `cylc` workflow with the test data, run:
+```bash
+cylc stop beaufort-sea-hudson-bay/*       # stops any currently running workflows
+cylc validate .                           # check the cylc configuration
+cylc install . -n beaufort-sea-hudson-bay # installs the current version of the workflow
+cylc play beaufort-sea-hudson-bay         # runs the workflow
+cylc tui beaufort-sea-hudson-bay          # opens the text user interface
```
Comment (cpaniaguam) [inline workflow/cylc/beaufort-sea-hudson-bay/README.md:8]: What do you think about adding, if possible, a utility that does all of this in one line?

File: workflow/cylc/sampled-examples/util/template.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import pandas
+import typer
+
+
+def main(datafile: pathlib.Path, index_col: str, index_val: str):
+    df = pandas.read_csv(datafile, index_col=index_col)
+    r = df.loc[index_val]
+    print(
+        f"""--icp {r.startdate} --fcp {r.enddate} --set BBOX="{r.left_x},{r.lower_y},{r.right_x},{r.top_y}" --set "LOCATION='{r.location}'" """
+    )
```
Comment (cpaniaguam) [inline workflow/cylc/sampled-examples/util/template.py:20]:
```suggestion
    template_with_good_name = (
    "--icp {startdate} "
    "--fcp {enddate} "
    '--set BBOX="{left_x},{lower_y},{right_x},{top_y}" '
    '--set "LOCATION=\'{location}\'"'
)

    filledin_template = template_with_good_name.format(**r.__dict__)

    print(filledin_template)

```

File: workflow/cylc/sampled-examples/flow.cylc [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            "${IMGDIR}/${IMGPREFIX}land.tiff" \
+            "${PREPROCESSDIR}"
+        """
+
+    [[preprocess-get-logs<satellite>]]
+        inherit = <satellite>
+        pre-script = """mkdir -p "${IMGDIR}/preprocess-logs" """
+        script = """
+            cp ${CYLC_WORKFLOW_RUN_DIR}/log/job/${CYLC_TASK_CYCLE_POINT}/preprocess_${SATELLITE}/NN/* ${IMGDIR}/preprocess-logs/.
+        """
```
Comment (cpaniaguam) [inline workflow/cylc/sampled-examples/flow.cylc:96]: Abstract long commands to short and descriptive names and reuse?

**Review by cpaniaguam — CHANGES_REQUESTED** [review]
Thanks for this! I left some suggestions to make the code a bit more readable.

File: workflow/cylc/beaufort-sea-hudson-bay/README.md (hunk 1 above)
Comment (hollandjg) [reply workflow/cylc/beaufort-sea-hudson-bay/README.md:8]: You could use `cylc vip <name>` for the middle three lines. (See https://cylc.github.io/cylc-doc/latest/html/user-guide/cheat-sheet.html for details.) 
I tend to just copy and paste the code into the command prompt, so whether it's three lines or five doesn't make much difference for me. For this example, I don't think it's worthwhile to shorten the code.

File: workflow/cylc/paper-domain/flow.cylc [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+    [[load<location, satellite>]]
+        inherit = <location><satellite>
+        script = """
+            mkdir -p "$IMGDIR"
+            BBOX=$(${FSDPROC} get-bbox "${CYLC_WORKFLOW_RUN_DIR}/locations.csv" "${LOCATION}")
+            ${FSDPROC} load ${IMGDIR}/${IMGPREFIX}-truecolor.tiff --kind truecolor --datetime ${DATE} --bbox "${BBOX}" --satellite "${SATELLITE}"
+            ${FSDPROC} load ${IMGDIR}/${IMGPREFIX}-cloud.tiff --kind cloud --datetime ${DATE} --bbox "${BBOX}" --satellite "${SATELLITE}"
+            ${FSDPROC} load ${IMGDIR}/${IMGPREFIX}-land.tiff --kind landmask --datetime ${DATE} --bbox "$BBOX"
+        """
```
[the remark this answers is not in the document]
Comment (hollandjg) [reply workflow/cylc/paper-domain/flow.cylc:65]: The name here is the task name `load<location, satellite>` which will resolve to `load_beaufort-sea_terra` in cylc. The idea is that the task names are `load<details>` and `preprocess<details>`. 
Do you have a suggestion for a more descriptive name? I don't want to use a long name as that will make the graph more unwieldy.

File: workflow/cylc/sampled-examples/README.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+## Cylc
+To run the `cylc` workflow with the test data, run:
+```bash
+cylc stop sampled-examples/*;
+cylc validate . &&
+cylc install . &&
+cylc play sampled-examples &&
+cylc tui sampled-examples 
+```
```
[the remark this answers is not in the document]
Comment (hollandjg) [reply workflow/cylc/sampled-examples/README.md:17]: Having a script would mean that we'd need to maintain an additional shell script which would need to be sourced in each new shell session, and maintained. I think if cylc provided a single command to do the stop and run that would be great.
For me it's not worth it to make that – I find it easy enough just to copy and paste this code, potentially using the `cylc vip` abbreviation discussed earlier.

File: workflow/cylc/sampled-examples/flow.cylc (hunk 3 above)
Comment (hollandjg) [reply workflow/cylc/sampled-examples/flow.cylc:96]: The name here is `preprocess-get-logs<satellite>` which might resolve to `preprocess-get-logs_terra` in cylc. Do you have a suggestion for a different name?

File: workflow/cylc/beaufort-sea-hudson-bay/README.md (hunk 1 above)
Comment (cpaniaguam) [reply workflow/cylc/beaufort-sea-hudson-bay/README.md:8]: I agree it's not worthwhile for one-off runs of the same sample workflow. I am looking at this from the perspective of a user running this very workflow for a different region later.  Don't examples illustrate general use? The user would need to make 4 updates (lines 1, 3, 4, and 5) to copied code if they needed to run this workflow for a different region. Just looking at it from a more abstract point of view.

File: workflow/cylc/paper-domain/flow.cylc (hunk 4 above)
[the remark this answers is not in the document]
Comment (cpaniaguam) [reply workflow/cylc/paper-domain/flow.cylc:65]: To elaborate a bit more, what I'm suggesting is to do something like this:
```suggestion
        PREPROC_LOC_SAT = """
            mkdir -p "$IMGDIR"
            BBOX=$(${FSDPROC} get-bbox "${CYLC_WORKFLOW_RUN_DIR}/locations.csv" "${LOCATION}")
            ${FSDPROC} load ${IMGDIR}/${IMGPREFIX}-truecolor.tiff --kind truecolor --datetime ${DATE} --bbox "${BBOX}" --satellite "${SATELLITE}"
            ${FSDPROC} load ${IMGDIR}/${IMGPREFIX}-cloud.tiff --kind cloud --datetime ${DATE} --bbox "${BBOX}" --satellite "${SATELLITE}"
            ${FSDPROC} load ${IMGDIR}/${IMGPREFIX}-land.tiff --kind landmask --datetime ${DATE} --bbox "$BBOX"
        """
        script = PREPROC_LOC_SAT 
```

Or maybe add a comment that'd serve a similar purpose.

File: workflow/cylc/sampled-examples/util/template.py (hunk 2 above)
Comment (hollandjg) [reply workflow/cylc/sampled-examples/util/template.py]: Thanks for this feedback!
I think that splitting the string across several lines is probably a good move. 
I think making the thing which does the template conversion into its own testable function might be a good compromise. I've updated this in a new version.

File: workflow/cylc/beaufort-sea-hudson-bay/README.md (hunk 1 above)
Comment (hollandjg) [reply workflow/cylc/beaufort-sea-hudson-bay/README.md]: That's a fair criticism. How about this? It's a bit more explicit, less repeaty.

File: workflow/cylc/beaufort-sea-hudson-bay/README.md [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+```
+
+or on one line:
+```bash
+cylc stop beaufort-sea-hudson-bay/* ; cylc validate . & cylc install . && cylc play beaufort-sea-hudson-bay && cylc tui beaufort-sea-hudson-bay
+```
```
[the remark this answers is not in the document]
Comment (hollandjg) [reply workflow/cylc/beaufort-sea-hudson-bay/README.md]: Removed.

File: workflow/cylc/sampled-examples/.gitignore
[the remark this answers is not in the document]
Comment (hollandjg) [reply workflow/cylc/sampled-examples/.gitignore:1]: Removed

File: workflow/cylc/sampled-examples/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (hollandjg) [reply workflow/cylc/sampled-examples/README.md]: I've updated all the code to use the `cylc vip` syntax instead.

File: workflow/cylc/sampled-examples/util/template.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    """
+    df = pandas.read_csv(datafile, index_col=index_column)
+    print(_template(df, row_index))
+
+
+def _template(df, row_index):
+    """
+    Examples:
+
+        >>> import io
```
Comment (cpaniaguam) [inline workflow/cylc/sampled-examples/util/template.py:27]:
```suggestion
        >>> import io, pandas
```

File: workflow/cylc/sampled-examples/util/template.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        >>> csv = '''id,location,center_lat,center_lon,top_left_lat,top_left_lon,lower_right_lat,lower_right_lon,left_x,right_x,lower_y,top_y,startdate,enddate
```
Comment (cpaniaguam) [inline workflow/cylc/sampled-examples/util/template.py:28]: For the purposes of the example, why not include the fields needed for the template only? It'd make for a more succinct illustration.

**Review by cpaniaguam — CHANGES_REQUESTED** [review]
Just two observations in the docstring example.

File: workflow/cylc/sampled-examples/util/template.py (hunk 8 above)
Comment (hollandjg) [reply workflow/cylc/sampled-examples/util/template.py]: That's an excellent idea, done.

File: workflow/cylc/sampled-examples/util/template.py (hunk 7 above)
Comment (hollandjg) [reply workflow/cylc/sampled-examples/util/template.py]: Good catch!

**Review by cpaniaguam — APPROVED** [review]
Thanks!

**Pull request merged**